maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,492 @@
+
+        self._cpu_reader.reset()
+
+    def _init_frame(self):
+        self._frame = build_frame(self._pm_amount, self.calc_max_snapshots())
+        self._snapshots = self._frame.snapshots
+
+    def step(self, tick: int):
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+        # All vm's cpu utilization at current tick.
+        cur_tick_cpu_utilization = self._cpu_reader.items(tick=tick)
+
+        # Process finished VMs.
+        self._process_finished_vm()
+        # Update all live VMs CPU utilization.
+        self._update_vm_workload(cur_tick_cpu_utilization=cur_tick_cpu_utilization)
+        # Update all PM CPU utilization.
+        self._update_pm_workload()
+
+        for vm in self._vm_item_picker.items(tick):
+            # TODO: Calculate
+            vm_info = VirtualMachine(
+                id=vm.vm_id,
+                cpu_cores_requirement=vm.vm_cpu_cores,
+                memory_requirement=vm.vm_memory,
+                lifetime=vm.vm_deleted - vm.timestamp + 1
+            )
+
+            if vm.vm_id not in cur_tick_cpu_utilization:
+                raise Exception(f"The VM id: '{vm.vm_id}' does not exist at this tick.")
+
+            vm_info.add_utilization(cpu_utilization=cur_tick_cpu_utilization[vm.vm_id])
+            vm_req_payload: VmRequestPayload = VmRequestPayload(
+                vm_info=vm_info,
+                remaining_buffer_time=self._buffer_time_budget
+            )
+            vm_request_event = self._event_buffer.gen_cascade_event(
+                tick=tick,
+                event_type=Events.REQUEST,
+                payload=vm_req_payload
+            )
+            self._event_buffer.insert_event(event=vm_request_event)
+            self._total_vm_requests += 1
+
+    def post_step(self, tick: int):
+        # Update energy to the environment metrices.
+        total_energy: float = 0.0
+        for pm in self._machines:
+            total_energy += pm.energy_consumption
+        self._total_energy_consumption += total_energy
+
+        if (tick + 1) % self._snapshot_resolution == 0:
+            # NOTE: We should use frame_index method to get correct index in snapshot list.
+            self._frame.take_snapshot(self.frame_index(tick))
+
+        # Stop current episode if we reach max tick.
+        return tick + 1 >= self._max_tick
+
+    def get_event_payload_detail(self) -> dict:
+        """dict: Event payload details of current scenario."""
+        return {
+            Events.REQUEST.name: VmRequestPayload.summary_key,
+            MaroEvents.PENDING_DECISION.name: DecisionPayload.summary_key
+        }
+
+    def get_agent_idx_list(self) -> List[int]:
+        """Get a list of agent index.
+
+        Returns:
+            list: List of agent index.
+        """
+        return [0]
+
+    def get_node_mapping(self) -> dict:
+        """dict: Node mapping."""
+        node_mapping = {}
```

Comment:
do we have a mapping?

Reply:
Currently no node mapping. The node "pms" id and index are the same. ( from 0 to ..)

Reply:
It's required by class ``Env`` (``Env.summary`` actually), but it is meaningless to this scenario?

Also, same to ``get_agent_idx_list()``, should we re-design the ``Env`` class or just implement a dummy interface in this scenario?

@ArthurJiang